### PR TITLE
Testplan WIP

### PR DESCRIPTION
**Pull request by GregAC**
~~Note the functional coverage plan referred to is still a PR: https://github.com/lowRISC/ibex/pull/1356 so the links are broken`~~

Functional coverage PR is merged to this has been updated to refer to it.

### CODE REVIEW

File: doc/03_reference/images/tb2.svg [hunk 1]
```diff
@@ -0,0 +1,989 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (vogelpi) [inline doc/03_reference/images/tb2.svg]: The figure is hardly readable in the html. If you have time, you can maybe reduce the left and right borders as well as the empty space on the right inside the violet box.

File: doc/03_reference/testplan.rst [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  One monitors the RVFI interface which provides details of retired instructions.
+  The other monitors relate to fetched instructions and instruction memory errors, more details are provided in :ref:`cosim`.
+  Additionally it connects to the monitor of the Memory Interface Agent for the instruction and data side via analysis ports.
+  The monitored transactions are used by a scoreboard to provide information to the co-simulation system allowing it to step the golden model and check its execution and memory activity against Ibex's behaviour.
+
+Memory Interface Agent
+  This provides a driver and a monitor for the :ref:`Ibex Memory Interface Protocol<lsu-protocol>`.
+  The driver provides fully randomised and configurable timings for responses and randomisation of error responses.
+  Two agents are instaniated, one for the data memory interface the other for the instruction memory interface.
+  Read data for memory responses is provided from a backing memory, writes requests update the contents of the backing memory.
```
Comment (vogelpi) [inline doc/03_reference/testplan.rst:50]: Nit: writes requests -> write requests

File: doc/03_reference/testplan.rst [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+Stimulus Strategy
+-----------------
+
+Stimulus falls into two categories:
+
+* Instructions to execute, these are generated by the `RISC-V DV random instruction <https://github.com/google/riscv-dv>`_ generator and provided to the testbench via a raw binary file.
+* Activity on external interfaces.
+
+Instructions are generated ahead of the test run so the test has no control over them.
+All external interfaces have their stimulus generated during the test run so can be controlled by the test.
```
Comment (vogelpi) [inline doc/03_reference/testplan.rst:78]: The combination of "test run" (the execution itself) and "test" (some form of test specification) may be confusing here. Also, the test actually specifies which instructions will be tested (e.g. bitmanip yes/no, wfi etc.), i.e., the test has an influence on the instructions, it's just offline.

Would it make sense to rephrase this as follows?
"Instructions are generated ahead of time so the test has no control over them at run time."
"All external interfaces have their stimulus generated at run time so can be controlled by the test."

**Review by vogelpi — APPROVED** [review]
LGTM @GregAC ! I have just minor comments, one on the differentiation between "test" and "test run". But I am no verification expert - if you decide to merge this as is it's completely fine with me.

File: doc/03_reference/testplan.rst [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+The aim is to check all possible externally observable behaviours of ``ibex_top`` against the golden model.
+The golden model used is the `Spike RISC-V ISS <https://github.com/riscv-software-src/riscv-isa-sim>`_.
+
+The testbench uses UVM.
+It consists of 3 agents:
+
+Co-simulation Agent:
+  This has multiple monitors.
+  One monitors the RVFI interface which provides details of retired instructions.
+  The other monitors relate to fetched instructions and instruction memory errors, more details are provided in :ref:`cosim`.
```
Comment (rswarbrick) [inline doc/03_reference/testplan.rst:42]: Nit: Should be semicolon before "more"

File: doc/03_reference/testplan.rst [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  Additionally it connects to the monitor of the Memory Interface Agent for the instruction and data side via analysis ports.
+  The monitored transactions are used by a scoreboard to provide information to the co-simulation system allowing it to step the golden model and check its execution and memory activity against Ibex's behaviour.
+
+Memory Interface Agent
+  This provides a driver and a monitor for the :ref:`Ibex Memory Interface Protocol<lsu-protocol>`.
+  The driver provides fully randomised and configurable timings for responses and randomisation of error responses.
+  Two agents are instaniated, one for the data memory interface the other for the instruction memory interface.
```
Comment (rswarbrick) [inline doc/03_reference/testplan.rst:49]: Nit: instantiated (and probably follow it with a colon, not a comma)

File: doc/03_reference/testplan.rst (hunk 2 above)
Comment (rswarbrick) [reply doc/03_reference/testplan.rst:50]: Also, probably a semicolon between the two halves.

File: doc/03_reference/testplan.rst [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  It provides randomised interrupt stimulus to Ibex when a test requests it.
+  Constraints can be used to control types of interrupts generated (e.g. NMI or not) and whether multiple interrupts should be raised together.
+
+Debug and reset signals are a single wire each so do not have a dedicated agent.
+Instead any sequence that wishes to use them will directly manipulate them via a virtual interface
+
+The testbench instantiates the agents described above along with the memory model used by both the data and instruction side memory agents.
+A test consists of executing a pre-built binary (which is loaded into the memory model at the start of the test via backdoor accesses) along with configuring agents to provide appropriate stimulus for the test.
+Some tests may use the agents to generate stimulus at particular times (e.g.  interrupts).
+A test may perform additional checking on top of the co-simulation golden model comparison where appropriate (e.g. ensuring a raised interrupt has cause an exception).
```
Comment (rswarbrick) [inline doc/03_reference/testplan.rst:67]: Nit: caused

File: doc/03_reference/testplan.rst [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+Stimulus Strategy
+-----------------
+
+Stimulus falls into two categories:
+
+* Instructions to execute, these are generated by the `RISC-V DV random instruction <https://github.com/google/riscv-dv>`_ generator and provided to the testbench via a raw binary file.
```
Comment (rswarbrick) [inline doc/03_reference/testplan.rst:74]: Nit: Probably a colon before "these"

File: doc/03_reference/testplan.rst [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+It is the responsibility of the regression run environment to ensure generated instructions are matched with appropriate tests (e.g. ensuring an exception handler is present where interrupts are expected).
+
+Stimulus generation will use a coverage based approach.
+Stimulus is developed based upon the :ref:`coverage-plan`.
+Where coverage is not being hit stimulus will be added to hit it.
+
+Tests
+-----
+
+As with stimulus test sequence development uses a coverage based approach.
```
Comment (rswarbrick) [inline doc/03_reference/testplan.rst:88]: I think you want a comma after "stimulus"? (I genuinely struggled to understand this sentence!)

File: doc/03_reference/testplan.rst [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+Tests will be added such that all coverage in the :ref:`coverage-plan` can be hit.
+All the details of specific tests will not be documented here.
```
Comment (rswarbrick) [inline doc/03_reference/testplan.rst:90]: Nit: This should probably be "Not all the details ... will be documented here."

File: doc/03_reference/testplan.rst [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+A test will execute a binary whilst running zero or more sequences that provide stimulus to external interfaces of ``ibex_top``.
+As the memory interfaces are all driven by Ibex, with any testbench generated activity in response to a request from Ibex, they do not require explicit sequences run by the test.
+Instead the test can configure the randomisation of memory delays as it wishes.
+Memory errors can be configured to always occur in statically defined areas of the memory map or a sequence can be used to inject them via the memory interface agent.
+
+The following sequences are available for tests to use.
+Each sequence is derived from a base sequence which provides controls to repeat the sequence at fixed or random internals, forever or after a random number of repeats.
+Full details can be found in `dv/uvm/core_ibex/tests/core_ibex_seq_lib.sv <https://github.com/lowRISC/ibex/blob/master/dv/uvm/core_ibex/tests/core_ibex_seq_lib.sv>`_.
+
+* ``irq_raise_seq`` - Raises one of more interrupts.
```
Comment (rswarbrick) [inline doc/03_reference/testplan.rst:102]: Nit: "one o**r** more" ?

**Review by rswarbrick — APPROVED** [review]
Looks good, modulo a couple of grammar nits.

**Comment by GregAC** [comment]
Thanks for the reviews @rswarbrick and @vogelpi I've addressed all your comments.

As part of making the testbench diagram more readable I've expanded the width of the documentation in general (from max of 800 to 1000), which is the other commit of the PR.

**Comment by vogelpi** [comment]
Very nice @GregAC , it's a good idea to change to the max width.
